src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java

```diff
@@ -149,26 +170,46 @@ public boolean requiresReference() {
     @Override
     public void onStartup() {
 
-        for (final String variantPath : variantFilePaths) {
+        List<VCFHeader> headers = new ArrayList<>(variantFilePaths.size());
 
-            final String variantURI = IOUtils.getPath(variantPath).toAbsolutePath().toUri().toString();
+        for (String variantFilePath : variantFilePaths) {
+
+            final String variantURI = IOUtils.getPath(variantFilePath).toAbsolutePath().toUri().toString();
             final Function<SeekableByteChannel, SeekableByteChannel> cloudWrapper = (cloudPrefetchBuffer > 0 ? is -> SeekableByteChannelPrefetcher.addPrefetcher(cloudPrefetchBuffer, is) : Function.identity());
             final Function<SeekableByteChannel, SeekableByteChannel> cloudIndexWrapper = (cloudIndexPrefetchBuffer > 0 ? is -> SeekableByteChannelPrefetcher.addPrefetcher(cloudIndexPrefetchBuffer, is) : Function.identity());
             final AbstractFeatureReader<VariantContext, LineIterator> reader = AbstractFeatureReader.getFeatureReader(variantURI, null, new VCFCodec(), true, cloudWrapper, cloudIndexWrapper);
 
+            VCFHeader header = (VCFHeader)reader.getHeader();
+
             // A GVCF file must contain only one sample, throw an exception otherwise
-            if (((VCFHeader)reader.getHeader()).getNGenotypeSamples() != 1) {
-                throw new UserException("Input GVCF: " + variantPath + " should contain data for one sample");
+            if (header.getNGenotypeSamples() != 1) {
+                throw new UserException("Input GVCF: " + variantFilePath + " should contain data for one sample");
+            }
+            String sampleName = header.getGenotypeSamples().get(0);
+            headers.add(header);
+            sampleNames.add(sampleName);
+            sampleNameToFileName.put(sampleName, variantFilePath);
+
+            // Import occurs in batches. By closing all readers here we release resources
+            // and preserve memory for later stages where readers are allocated
+            // for samples in a given batch
+            try {
+                reader.close();
+            } catch (IOException e) {
+                throw new GATKException("FeatureReader close() failed for " + sampleName, e);
             }
-
-            final String sampleName = ((VCFHeader) reader.getHeader()).getGenotypeSamples().get(0);
-            sampleToVCFMap.put(sampleName, reader);
         }
 
-        createMergedHeader();
+        mergedHeader = VCFUtils.smartMergeHeaders(headers, true);
```

Comment:
Reading all the headers up-front like this, accumulating them in a collection, and then merging them might prove way too expensive over GCS. In a future PR (does not need to happen in this PR!), we might want to change this. One possibility is that we could pass in (via a new argument) a single input whose header to treat as canonical. I'll file a ticket to explore this.